share/wake/lib/system/http.wake

```diff
@@ -175,53 +165,106 @@ def parseCurlResponse (str: String): Result HttpResponse Error =
     else failWithError "Request Body and FormData are both set. This is not allowed."
 
-    def cmd =
-        (
-            Some ("curl", Nil),
-            Some ("-i", Nil), # Include response headers
-            Some ("--request", method | methodToString, Nil),
-            Some ("--url", url, Nil),
-            Some (headers | mapFlat headerToCurlFlag),
-            (body | omap bodyToCurlFlag),
-            (formData | omap (mapFlat formDataToCurlFlag)),
-        )
-        | mapPartial (_)
-        | flatten
+    # Curl does a preflight check bofore sending the request if it is large. In core use cases
+    # this requres a clunky retry of the request. For now we accpt the tradeoff of wasting time
+    # uploading a reject file to not muddy the api. Setting 'Expect' to empty disables the check.
+    # See https://everything.curl.dev/http/post/expect100 for more details.
+    def heads = (HttpHeader "Expect" ""), headers
+
+    (
+        "curl",
+        "--request",
+        (method | methodToString),
+        "--url",
+        url,
+        (heads | mapFlat headerToCurlFlag) ++
+        (body | omap bodyToCurlFlag | optionToList | flatten) ++
+        (formData | omap (mapFlat formDataToCurlFlag) | optionToList | flatten) ++
+        extraFlags
+    )
+    | Pass
+
+# Makes the request specified by request and parses into a HttpResponse.
+#
+# The default method to make a request. It should be used when the response is string encodable.
+# WARNING: This function will likely break the sandbox and make your build unreliable
+export def makeRequest (request: HttpRequest): Result HttpResponse Error =
+    # -i -> Include response headers
+    require Pass cmd = makeCurlCmd request ("-i", Nil)
+
+    def method = request.getHttpRequestMethod
+    def url = request.getHttpRequestUrl
 
     require Pass stdout =
         makeExecPlan cmd Nil
         | setPlanLabel "http: {method | methodToString} {url}"
-        | setPlanStdout logInfo
+        | setPlanStdout logNever
         # Curl dumps download stats to stderr
         | setPlanStderr logInfo
         # Web is unreliable and should never be reused
         | setPlanPersistence ReRun
+        | setPlanFnOutputs (\_ Nil)
         | runJobWith localRunner
+        | setJobTag "http" "{method | methodToString} {url}"
         | getJobStdout
 
     parseCurlResponse stdout
+
+# Makes the request specified by request and writes body to a file.
+#
+# The backup method to make a request. It must be used when the response is not string encodable.
+# WARNING: This function will likely break the sandbox and make your build unreliable
+export target makeBinaryRequest (request: HttpRequest): Result Path Error =
+    # We can't create a unique file name ahead of time by hashing the contents so we hash the
+    # request instead. This won't create 'file output by many jobs' in a single invocation since
+    # the function is a target, and across multiple invocations its fine as the job is set to
```

Comment:
The other consideration with making the HTTP stuff general is that someone might not give this a GET `HttpRequest` -- by making it a `target`, if we have something like a POST (which is explicitly non-idempotent, as opposed to PUT) then someone might *expect* to be submitting multiple identical requests to their server but the server will only see the first.  I'd question why they're wanting their build system to be posting keepalives or whatever, but I can foresee someone at some point in some project adding visualization tooling which is poorly-enough written to not be able to handle being left alone even while nothing is changing.

Reply:
hmmm yeah. That is a concern. I'd really prefer to not specialize the API but this _will_ confuse someone eventually.

On one hand any non-protobuf POST request should never be returning binary data (and a protobuf POST would need a whole other library anyways) but on the other this will come up.

Personally I'm tempted to just do something like 

```
def _ = 
    if method !=~ "GET" then
        printlnLogLevel "makeBinaryRequest is idenpoent and using non-GET verbs will act unexpectedly" logWarning
    else Unit
```

since very few people will use the not recommended sub api in the not recommended api

Reply:
Yeah, I think a warning is probably sufficient here.  Anybody using this (should) already know they have to be particularly careful, so some degree of sub-par behaviour in particular edge cases should be fine.

Thinking about it, it does mean the `ReRun` we *say* the job is using actually functions like `Once`, but I don't think that's a distinction we have to make in the code.

Reply:
yeah the target does actually just bottom out in Once (which also uses a more precise target). I'll just make that change